chiselFrontend/src/main/scala/chisel3/core/Module.scala

```diff
@@ -25,16 +25,22 @@ object Module {
     // module de-duplication in FIRRTL emission.
     val childSourceInfo = UnlocatableSourceInfo
 
-    val parent: Option[Module] = Builder.currentModule
-    val m = bc.setRefs() // This will set currentModule!
+    val m = bc.setRefs() // This will invoke bc which pushes it to the top of the Module stack!!!
     m._commands.prepend(DefInvalid(childSourceInfo, m.io.ref)) // init module outputs
-    Builder.currentModule = parent // Back to parent!
+    // Since the above only calls the constructor for the Module once, if you double wrap then we
+    // end up popping twice. If you forget to Module wrap, then a outter Module apply will trigger this
+    if (Builder.moduleStack.isEmpty || Builder.moduleStack.head != m) {
+      throwException("Error: Top of Module stack is not the Module expected. " +
+                     "You probably forgot a Module() wrapper or double Module() wrapped.")
+
+    }
```

Comment:
I suggest splitting these cases apart and giving the more precise error message in either case.